### ISSUE TITLE
Flaky OOM on SnapshotTransactionTest

### ISSUE THREAD

**Issue by jeremyk-91** [palantir/atlasdb#1392]
`SnapshotTransactionTest` sometimes fails with an OOM on external CircleCI.

See [flake 1](https://circleci.com/gh/palantir/atlasdb/5513), [flake 2](https://circleci.com/gh/palantir/atlasdb/5558), among others.

I have not seen this occur on internal CI.

Possible hypotheses:
1. Memory leaks in existing tests (possibly `StreamTest` which runs just before it and is relatively new, or `SnapshotTransactionTest`).
2. Recent changes made to speed up the build have caused additional memory pressure. Although steps have been taken to mitigate this it appears builds are still flaky.

(Possibly both as well, of course.)

ETA: It would probably be a good idea to add GC logging to our builds, to help troubleshoot memory-related issues more easily. I had to stick the relevant flags in our circle.yml and rerun the builds until specifically this flake reproduced itself, which was a bit unpleasant. (Note: There are other flakes where the SnapshotTransactionTest deadlocks (no output for 300s), or one of the lock tests hangs!)

```
[00:09:07] com.palantir.atlasdb.transaction.impl.SnapshotTransactionTest > testConcurrentWriteWriteConflicts STARTED
[00:13:57] Unexpected exception thrown.
[00:13:57] java.lang.OutOfMemoryError: GC overhead limit exceeded
[00:13:57] 	at com.sun.proxy.$Proxy88.output(Unknown Source)
[00:13:57] 	at sun.reflect.GeneratedMethodAccessor718.invoke(Unknown Source)
[00:13:57] 	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[00:13:57] 	at java.lang.reflect.Method.invoke(Method.java:497)
[00:13:57] 	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:35)
[00:13:57] 	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
[00:13:57] 	at org.gradle.internal.remote.internal.hub.MessageHub$Handler.run(MessageHub.java:377)
[00:13:57] 	at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:54)
[00:13:57] 	at org.gradle.internal.concurrent.StoppableExecutorImpl$1.run(StoppableExecutorImpl.java:40)
[00:13:57] 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
[00:13:57] 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
[00:13:57] 	at java.lang.Thread.run(Thread.java:745)
[00:13:57] Could not read error output of: Gradle Test Executor 5.
[00:13:57] java.lang.OutOfMemoryError: GC overhead limit exceeded

**Comment by jeremyk-91** [comment]
`StreamTest` is probably not causing a memory leak. Running with `-verbosegc`:

```
// BeforeClass
[GC (System.gc())  57736K->3798K(251392K), 0.0040767 secs]
[Full GC (System.gc())  3798K->3649K(251392K), 0.0104692 secs]

// AfterClass
[GC (System.gc())  94257K->60631K(699392K), 0.0037621 secs]
[Full GC (System.gc())  60631K->3300K(699392K), 0.0312149 secs]
```

I'll run a few tests on circle with `-verbosegc`, to generate some hopefully more interesting data.

**Comment by schlosna** [comment]
@jeremyk-91 do we have any hprof from these tests? We should make sure tests are run with `-XX:+HeapDumpOnOutOfMemoryError` if possible and pull any *.hprof into test results.

If possible, would be good to enable GC logging for tests in general, with something along the lines of `-verbose:gc -XX:+PrintGC -XX:+PrintGCDateStamps -XX:+PrintGCDetails -XX:-TraceClassUnloading -Xloggc:$(date +%Y%m%d%H%M%S).gc.log`

**Comment by jeremyk-91** [comment]
@schlosna Unfortunately we don't normally run the tests with that (it seems we normally run with only `-Xmx512M`), though I've been rerunning the tests [with it](https://github.com/palantir/atlasdb/commit/27e89b0fd6c42911fc92500c050de5924b79fc84
) - currently awaiting results.

Yep, would be good to add the rest of the GC logging as well, and to leave this in since we do know there is memory pressure.

**Comment by jeremyk-91** [comment]
Mm, heapdump from [latest flake](https://circleci.com/gh/palantir/atlasdb/5580#artifacts/containers/5) shows it's actually _Gradle_ which OOMed, not our tests. The dump shows a *lot* of char[]s, which seem to include test outputs as well as some of our Gradle files.

Will look at this again after dinner.

**Comment by jeremyk-91** [comment]
<img width="507" alt="screen shot 2017-01-03 at 19 08 49" src="https://cloud.githubusercontent.com/assets/5955510/21619566/25ebee94-d1e8-11e6-97ed-39cfbb149d20.png">
<img width="1217" alt="screen shot 2017-01-03 at 19 07 51" src="https://cloud.githubusercontent.com/assets/5955510/21619565/25ea1484-d1e8-11e6-8a06-005d0f212a4d.png">

So it seems that all / many of the log lines we write are kept in memory, as are our Gradle source files (the top line is from `TransactionConflictException` after being barraged with RangeRequests). There are 2 possible ideas to fix this:

1. See if it's safe to bump `-Xmx` a bit higher.
2. Set Gradle to fork a new test VM after some number of tests have run (`forkEvery`).

Trying 1 first. But we'll see.

**Comment by jeremyk-91** [comment]
Further update post discussion with @schlosna:

* This is probably caused by Gradle internally buffering log messages and not disposing them. 
  - ~25% heap is retained by `org.gradle.internal.remote.internal.hub.queue.MultiEndPointQueue`
  - `select count(heap.objects('java.lang.String'), "s.toString().startsWith('18:')")` in jvisualvm returns `775604`, which is very close to our `776271` total
  - We considered the possibility of this being an [SLF4J memory leak](http://jira.qos.ch/browse/SLF4J-364), but we use Gradle 3.2.1 which uses SLF4J 1.7.21, so this is not the issue.
* A good fix could be to reduce the quantity of logging, at least when running the tests (ideally, when running the tests on external circle).

**Comment by jeremyk-91** [comment]
Fixed by #1395.
